### PR TITLE
Fix SQL query gen for multiple _type params

### PR DESCRIPTION
**Pull request by namadabu**
## Description
As mentioned in the issue, in some situations we generate an incorrect query when paging through results for queries that contain multiple `_type` parameters. This happens when we are returning the results for the "last" resource type. An example of the query we generate:

```
SELECT DISTINCT TOP (@p0) r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(1 AS bit) AS IsMatch, CAST(0 AS bit) AS IsPartial, r.IsRawResourceMetaSet, r.SearchParamHash, r.RawResource
      FROM dbo.Resource r
      WHERE (
              ResourceTypeId = @p1 AND ResourceSurrogateId > @p2
              OR ResourceTypeId IN ()
          )
          AND IsHistory = 0
          AND IsDeleted = 0
      ORDER BY r.ResourceTypeId ASC, r.ResourceSurrogateId ASC
```

The extra `()` throws an exception since that is not a valid query. The `OR` clause is only required when there are other resource types to search for. Since we are only returning the resources of the "last" resource type (represented by p2) we don't need the `OR` clause. 

Updated the query generator logic to add the `OR` clause only when there are other resource types to search for.

## Related issues
Addresses #1985 

## Testing
Added new tests. Manual tests on local deployment.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs [hunk 1]
```diff
@@ -192,6 +164,68 @@ public async Task GivenVariousTypesOfResources_WhenSearchingAcrossAllResourceTyp
...
+            {
+                expectedResources.Add(patient);
+            }
+
+            foreach (Observation observation in observations)
+            {
+                expectedResources.Add(observation);
+            }
+
+            await ExecuteAndValidateBundle($"?_type=Patient,Observation&_tag={tag}&_count=3", sort: false, pageSize: 3, expectedResources.ToArray());
```
Comment (Ivanidzo4ka) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs:226]: >ExecuteAndValidateBundle [](http://example.com/codeflow?start=18&length=24)

nit: can we sort by lastupdated and have one test?

Comment (namadabu) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs:226]: I had two separate tests because I wanted to highlight the difference in behavior between Cosmos and SQL for such queries.

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs [hunk 2]
```diff
@@ -49,24 +49,11 @@ public async Task GivenResourceWithVariousValues_WhenSearchedWithMultipleParams_
         {
             // Create various resources.
             Patient[] patients = await Client.CreateResourcesAsync<Patient>(
-                p => SetPatientInfo(p, "Seattle", "Robinson"),
-                p => SetPatientInfo(p, "Portland", "Williamas"),
-                p => SetPatientInfo(p, "Seattle", "Jones"));
+                p => SetPatientInfo(p, "Seattle", "Robinson", tag: null),
+                p => SetPatientInfo(p, "Portland", "Williamas", tag: null),
+                p => SetPatientInfo(p, "Seattle", "Jones", tag: null));
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs:54]: Should we add a tag to these tests? I know that not having tags has caused flaky tests.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs [hunk 3]
```diff
@@ -77,30 +64,15 @@ public async Task GivenResourceWithVariousValues_WhenSearchedWithCityParam_ThenO
...
-                p => SetPatientInfo(p, "Amsterdam", "Hill"));
+                p => SetPatientInfo(p, "seattle", "Robinson", tag), // match
+                p => SetPatientInfo(p, "Portland", "Williamas", tag),
+                p => SetPatientInfo(p, "SEATTLE", "Skouras", tag), // match
+                p => SetPatientInfo(p, "Sea", "Luecke", tag),
+                p => SetPatientInfo(p, "Seattle", "Jones", tag), // match
+                p => SetPatientInfo(p, "New York", "Cook", tag),
+                p => SetPatientInfo(p, "Amsterdam", "Hill", tag));
 
             await ExecuteAndValidateBundle("Patient?address-city=Seattle", patients[0], patients[2], patients[4]);
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/BasicSearchTests.cs:75]: Not part of this PR, but this test defines a tag but doesn't use it in its search.

**Review by LTA-Thinking — APPROVED** [review]
